### PR TITLE
[LSP] Semantic tokens: Utilize frozen partial compilations

### PR DESCRIPTION
**Pull request by allisonchou**
Allows Roslyn's LSP semantic token handlers to utilize frozen compilations to compute tokens. The switch to using frozen compilations is fairly straightforward, however Razor needs to know if we're returning partial tokens or not so they know whether to keep asking us for tokens. To accomplish this, we return a `IsFinalized` property along with our results that Razor can deserialize.

Also modified the handlers' logic to not cache 0-length tokens as it is unnecessary.

Likely should be dual-inserted with the Razor-side PR: https://github.com/dotnet/aspnetcore-tooling/pull/4156

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/RoslynSemanticTokens.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        /// <summary>
+        /// True if the token set may be incomplete.
+        /// </summary>
+        /// <remarks>
+        /// Certain clients such as Razor need to know whether we're returning partial results.
+        /// This may occur if the full compilation is not yet available.
+        /// </remarks>
+        [DataMember(Name = "isPartial")]
+        public bool IsPartial { get; set; }
```
[the remark this answers is not in the document]
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/RoslynSemanticTokens.cs]: `IsFinalized` fits well to me, good suggestion! `IsIncomplete` seems too similar to `IsPartial` imo.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 2]
```diff
@@ -116,9 +117,15 @@ internal class SemanticTokensHelpers
...
             // can pass in a range if they wish.
-            var textSpan = range == null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
+            var textSpan = range is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
 
-            var classifiedSpans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            // If the full compilation is not yet available, we'll try getting a partial one.
+            var frozenDocument = document.WithFrozenPartialSemantics(cancellationToken);
+            var semanticModel = await frozenDocument.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(semanticModel);
+            var isPartial = !(document.Project.TryGetCompilation(out var compilation) && compilation == semanticModel.Compilation);
```
[the remark this answers is not in the document]
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs]: Gotcha, I updated the code so we just directly do `document = document.WithFrozenPartialSemantics(cancellationToken)`, so it should hopefully now prevent the problem you mentioned

**Comment by allisonchou** [comment]
#55910 has been merged, which should clear this PR.

I spent quite a bit of time today trying to figure out how feasible adding a test for partial semantics would be + checked with Jason, and in summary it seems more difficult than expected. 😕 It appears the test infrastructure by default gets the full compilation and trying to figure out how to get a weak compilation in the LSP layer will probably involve some time. The test Jason added is in a different layer than LSP and the constructs it uses aren't available in the LSP layer currently. 

Considering Cyrus' PR to add frozen semantics to non-LSP scenarios [didn't include tests either](https://github.com/dotnet/roslyn/pull/51704/files#diff-ddf5793d3c96527cf4835c68a8ad16718cec4df3b8c4e7aa2d6b1c979340e4f4R50), I'm wondering if we can table this for the future (potentially P5) so we can include this in time for P4. Thoughts @ryanbrandenburg @dibarbet? I can file an issue to add tests if we decide to go this route. I tested manually and confirmed things work, although I can def see how not having tests is risky.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs [hunk 3]
```diff
@@ -64,28 +64,42 @@ public SemanticTokensEditsHandler(SemanticTokensCache tokensCache)
...
             // If we have edits, generate a new ResultId. Otherwise, re-use the previous one.
             if (editArray.Length != 0)
             {
                 resultId = _tokensCache.GetNextResultId();
-                var updatedTokens = new LSP.SemanticTokens { ResultId = resultId, Data = newSemanticTokensData };
-                await _tokensCache.UpdateCacheAsync(
-                    request.TextDocument.Uri, updatedTokens, cancellationToken).ConfigureAwait(false);
+                if (newSemanticTokensData.Length > 0)
+                {
+                    var updatedTokens = new LSP.SemanticTokens { ResultId = resultId, Data = newSemanticTokensData };
```
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:92]: Should this be `RoslynSemanticTokens`, for consistency if nothing else?

**Review by davidwengier — APPROVED** [review]

**Review by dibarbet — APPROVED** [review]

**Review by ryanbrandenburg — APPROVED** [review]

**Comment by allisonchou** [comment]
Merging. Filed https://github.com/dotnet/roslyn/issues/56038 to add a test in the future.

**Pull request merged**